# 2. Data Tables in Python

In [1]:
import pandas as pd
df = pd.DataFrame({

'StudentID' : [264422,264423,264444,264445,264446], # StudentID column and its data

'FirstName' : ['Steven','Alex','Bill','Mark','Bob'], # FirstName column and its data

'EnrolYear' : [2010,2010,2011,2011,2013], # Enrolment year column and its data

'Math' : [100,90,90,40,60], # Mathematics score column and its data

'English' : [60,70,80,80,60] # English score column and its data

})
df

English  EnrolYear FirstName  Math  StudentID
0       60       2010    Steven   100     264422
1       70       2010      Alex    90     264423
2       80       2011      Bill    90     264444
3       80       2011      Mark    40     264445
4       60       2013       Bob    60     264446

#### Practice: 
Create another data table called "df2" that contains the height of some of the students. It should contain the following data: 



In [2]:
df2 = pd.DataFrame({
'StudentID' : [264422,264423,264444,264445], 
'Height' : [160,155,175,175], 
})
df2

Height  StudentID
0     160     264422
1     155     264423
2     175     264444
3     175     264445

In [3]:
df['FirstName']
df.FirstName
df[df['FirstName'] == 'Alex']


English  EnrolYear FirstName  Math  StudentID
1       70       2010      Alex    90     264423

#### Practice : 
Show the details of the students who enrolled in the year 2011?


In [4]:
df[df.EnrolYear==2011]

English  EnrolYear FirstName  Math  StudentID
2       80       2011      Bill    90     264444
3       80       2011      Mark    40     264445

In [5]:
filt = (df.FirstName != 'Bob') & (df.StudentID > 264423)
df[filt]

English  EnrolYear FirstName  Math  StudentID
2       80       2011      Bill    90     264444
3       80       2011      Mark    40     264445

In [6]:
# Add new column
df['Total'] = df['Math'] + df['English']
df

English  EnrolYear FirstName  Math  StudentID  Total
0       60       2010    Steven   100     264422    160
1       70       2010      Alex    90     264423    160
2       80       2011      Bill    90     264444    170
3       80       2011      Mark    40     264445    120
4       60       2013       Bob    60     264446    120

#### Practice :
Add a new column showing the average mark of "Math" and "English" marks for each student. (Average = Total/number of subjects)


In [7]:
df['AvgGrades']=(df.English+df.Math)/2
#you can use the newly generated column, Total, as well.
# df['AvgGrades']=dfTotal./2
df

English  EnrolYear FirstName  Math  StudentID  Total  AvgGrades
0       60       2010    Steven   100     264422    160       80.0
1       70       2010      Alex    90     264423    160       80.0
2       80       2011      Bill    90     264444    170       85.0
3       80       2011      Mark    40     264445    120       60.0
4       60       2013       Bob    60     264446    120       60.0

#### Practice :
Practice 5: Write a filter to select students who have scored more than 150 in total and who have achieved a subject score (in Maths or English) of 90 or more.

In [8]:
df[(df.Total>150) & ((df.English>=90) |( df.Math>=90))]

English  EnrolYear FirstName  Math  StudentID  Total  AvgGrades
0       60       2010    Steven   100     264422    160       80.0
1       70       2010      Alex    90     264423    160       80.0
2       80       2011      Bill    90     264444    170       85.0

 In our current table we have two columns "Math" and "English", both of which contain student marks. We can split the information for each student into different rows using the "melt" command, as follows:

In [9]:
df = pd.melt(df, id_vars=['EnrolYear','FirstName','StudentID'],value_vars=['Math','English'],var_name='Subject')

In [10]:
df

EnrolYear FirstName  StudentID  Subject  value
0       2010    Steven     264422     Math    100
1       2010      Alex     264423     Math     90
2       2011      Bill     264444     Math     90
3       2011      Mark     264445     Math     40
4       2013       Bob     264446     Math     60
5       2010    Steven     264422  English     60
6       2010      Alex     264423  English     70
7       2011      Bill     264444  English     80
8       2011      Mark     264445  English     80
9       2013       Bob     264446  English     60

In [11]:
df.rename(columns = {'value':'Score'}, inplace = True)
df

EnrolYear FirstName  StudentID  Subject  Score
0       2010    Steven     264422     Math    100
1       2010      Alex     264423     Math     90
2       2011      Bill     264444     Math     90
3       2011      Mark     264445     Math     40
4       2013       Bob     264446     Math     60
5       2010    Steven     264422  English     60
6       2010      Alex     264423  English     70
7       2011      Bill     264444  English     80
8       2011      Mark     264445  English     80
9       2013       Bob     264446  English     60

In [12]:
df3 = pd.merge(df, df2, on=['StudentID'])
df3

EnrolYear FirstName  StudentID  Subject  Score  Height
0       2010    Steven     264422     Math    100     160
1       2010    Steven     264422  English     60     160
2       2010      Alex     264423     Math     90     155
3       2010      Alex     264423  English     70     155
4       2011      Bill     264444     Math     90     175
5       2011      Bill     264444  English     80     175
6       2011      Mark     264445     Math     40     175
7       2011      Mark     264445  English     80     175

#### Practice:
What is the difference that can you see after merging to datasets?
Print out the merged table. Based on the output, explain what you understand about the parameter on=['StudentID'].

In [13]:
print(df)
print(df2)

   EnrolYear FirstName  StudentID  Subject  Score
0       2010    Steven     264422     Math    100
1       2010      Alex     264423     Math     90
2       2011      Bill     264444     Math     90
3       2011      Mark     264445     Math     40
4       2013       Bob     264446     Math     60
5       2010    Steven     264422  English     60
6       2010      Alex     264423  English     70
7       2011      Bill     264444  English     80
8       2011      Mark     264445  English     80
9       2013       Bob     264446  English     60
   Height  StudentID
0     160     264422
1     155     264423
2     175     264444
3     175     264445


Student ID 264446 does not exist in df3 as the ID does not exist in one of the data frames(df2).

#### Practice : 
Looking closely at the data in the resulting DataFrame, is there a student is missing after the merge? How can you display all students?

In [14]:
df4 = pd.merge(df, df2, on=['StudentID'],how='left')
df4

EnrolYear FirstName  StudentID  Subject  Score  Height
0       2010    Steven     264422     Math    100   160.0
1       2010      Alex     264423     Math     90   155.0
2       2011      Bill     264444     Math     90   175.0
3       2011      Mark     264445     Math     40   175.0
4       2013       Bob     264446     Math     60     NaN
5       2010    Steven     264422  English     60   160.0
6       2010      Alex     264423  English     70   155.0
7       2011      Bill     264444  English     80   175.0
8       2011      Mark     264445  English     80   175.0
9       2013       Bob     264446  English     60     NaN

### Reading CSV and Excel files into DataFrames

In [15]:
ufo_reports = pd.read_csv('uforeports.csv')
ufo_reports.head() # print the first 5 records 


City Colors Reported Shape Reported State  \
0                Ithaca             NaN       TRIANGLE    NY   
1           Willingboro             NaN          OTHER    NJ   
2               Holyoke             NaN           OVAL    CO   
3               Abilene             NaN           DISK    KS   
4  New York Worlds Fair             NaN          LIGHT    NY   

                  Time  
0  1930-01-06 22:00:00  
1  1930-06-30 20:00:00  
2  1931-02-15 14:00:00  
3  1931-01-06 13:00:00  
4  1933-04-18 19:00:00

#### Practice : 
How can we display the last 5 records? How can we display the last record only? 



In [16]:
ufo_reports.tail(5)

City Colors Reported Shape Reported State                 Time
18236   Grant Park             NaN       TRIANGLE    IL  2000-12-31 23:00:00
18237  Spirit Lake             NaN           DISK    IA  2000-12-31 23:00:00
18238  Eagle River             NaN            NaN    WI  2000-12-31 23:45:00
18239  Eagle River             RED          LIGHT    WI  2000-12-31 23:45:00
18240         Ybor             NaN           OVAL    FL  2000-12-31 23:59:00

In [17]:
ufo_reports.tail(1)

City Colors Reported Shape Reported State                 Time
18240  Ybor             NaN           OVAL    FL  2000-12-31 23:59:00

In [18]:
# read excel
ufo_reports_xls = pd.read_excel('uforeports_excel.xls', sheet_name='uforeports')

In [19]:
print(ufo_reports.Time.dtypes)

object


In [20]:
print(ufo_reports_xls.Time.dtypes)

datetime64[ns]


Notice that Time columns in two DataFrames are indeed different. In the table read in from the CSV file, the Time column is just an object, while for the XLS file it was given a datetime format. We can change the format so they are consistent:

In [21]:
ufo_reports.Time = pd.to_datetime(ufo_reports.Time) # change to datetime format
print (ufo_reports.Time.dtypes) # print its datatype now

datetime64[ns]


In [22]:
ufo_reports.head() # check whether it looks the same as the other dataframe

City Colors Reported Shape Reported State  \
0                Ithaca             NaN       TRIANGLE    NY   
1           Willingboro             NaN          OTHER    NJ   
2               Holyoke             NaN           OVAL    CO   
3               Abilene             NaN           DISK    KS   
4  New York Worlds Fair             NaN          LIGHT    NY   

                 Time  
0 1930-01-06 22:00:00  
1 1930-06-30 20:00:00  
2 1931-02-15 14:00:00  
3 1931-01-06 13:00:00  
4 1933-04-18 19:00:00

#### Practice:
Write a statement to read in titanic data (titanic.csv) into a pandas DataFrame called ‘titanic’. Then print out the first few rows.

In [23]:
titanic=pd.read_csv('titanic.csv')
titanic.head(5)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [24]:
print(titanic.shape)

(891, 15)


How many rows and columns are there? 
891 rows and 15 cols


In [25]:
titanic.dtypes

survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
who             object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object

In [26]:
titanic.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [27]:
titanic.age.value_counts().head()

24.0    30
22.0    27
18.0    26
19.0    25
30.0    25
Name: age, dtype: int64

# Group By

In [28]:
titanic['age'].mean()
titanic['age'].max() 
titanic['age'].min() 
titanic['age'].sum() # total years
titanic['age'].std() # standard deviation
titanic['age'].median() # half of the people were older (/younger) than ...
titanic['age'].mode() # most common age

0    24.0
dtype: float64

In [29]:
sex_class = titanic.groupby(['sex','class'])['age']
sex_class.mean()

sex     class 
female  First     34.611765
        Second    28.722973
        Third     21.750000
male    First     41.281386
        Second    30.740707
        Third     26.507589
Name: age, dtype: float64

#### Practice:
Interpret the output, e.g. what did you notice about the average age of the Titanic’s passengers in regards to their classes? How about the relationship between age and gender?
For both gender, the mean age increase for the higher (rank) class. Male has higher mean age than the female, for all classes.

#### Practice:
Use other aggregation functions and groupings to determine which class had the oldest and youngest passengers and which gender had the largest amount of variation in age (standard deviation).

In [67]:
sex_class.min()

sex     class 
female  First     2.00
        Second    2.00
        Third     0.75
male    First     0.92
        Second    0.67
        Third     0.42
Name: age, dtype: float64

In [68]:
sex_class.max()

sex     class 
female  First     63.0
        Second    57.0
        Third     63.0
male    First     80.0
        Second    70.0
        Third     74.0
Name: age, dtype: float64

In [69]:
sex_class.std()

sex     class 
female  First     13.612052
        Second    12.872702
        Third     12.729964
male    First     15.139570
        Second    14.793894
        Third     12.159514
Name: age, dtype: float64

In [34]:
#If you want to repeat the group by without gender, you can use following code and apply the above functions again
sex_class1 = titanic.groupby(['class'])['age']

class
First     0.92
Second    0.67
Third     0.42
Name: age, dtype: float64